### ISSUE TITLE
[BUG] Various bug regarding damage calculation (Previously discussing BB damage)

### ISSUE THREAD

**Issue by nattwara** [GMBoreas/trocalculator#251]
Bowling bash damage with bow type weapon is bugged.

## IGNORE THIS COMMENT, THE THEORY WITH ELEMENT IS NOT THE CAUSE OF THE BUG. PLEASE REFER TO OTHER COMMENTS BELOW ##

The damage calculation take property of both bow and arrow and calculated twice.
Which is not supposed to happen as arrow property should override bow property.

Calculator test sample:
Player: Rogue, base 65 job 1, 99+0 dex (base 155 atk)
Target: Injustice (Undead race, Shadow 2 Property)
100% neutral
75% earth, fire, water, wind
25% poison
150% holy
-25% shadow
50% ghost
0% undead

Result from calculator:
Rudra Bow rental (185 weapon atk, holy property, 99+0 dex, 365 base atk)
-Neutral property arrow: ~4.7k
-Silver arrow: ~7.1k
-Neutral arrow + select earth property manually: ~3.5k
-Silver arrow + select earth property manually: ~3.5k
Gakkung Bow [Incantation Samurai, Andre] (100+65+20 atk, neutral property, 99+0 dex, 365 base atk)
-arrow: ~3.6k
-silver ~5.5k
-arrow + select earth property manually: ~2.7k
-silver + select earth property manually: ~2.7k
Arbalest [Inca, Zipper Bear] (90+65+30 atk, neutral property, 97+2 dex, 365 base atk)
-arrow: ~3.6k
-silver ~5.5k
-arrow + select earth property manually: ~2.7k
-silver + select earth property manually: ~2.7k

From ~3.6k to ~4.7k is about 33% increase for no reason with doesn't happen in-game.
The only different is the rental rudra itself is holy property.
It could force attack to be holy with neutral arrow, will be testing in-game for confirmation.
But with silver arrow, the arrow should force the attack to be holy only once. which should result in about 5.5k like the other bow.

I tried calculating manually seperating the bow atk and atk that did not come from the bow:
185 x 500% * 2 * 150%(holy) = 2775
(365-185) x 500% * 2 = 1800
2775 + 1800 = 4575, closed to rental rudra + neutral arrow.
With silver arrow, 4575 = 6862.5, which came close to what happen in the calculator and which mean the atk from the bow got x150% twice, while the other 180 atk only x150% once.
This shouldn't be the case because arrow are supposed to override bow and the multiply happen only once.
(apart from the non-neutral bow + neutral arrow that I'm uncertain at the moment)

Also there's a report on forum from 2017,

https://forum.talonro.com/topic/24117-bowling-bash-damage-with-a-bow-is-inaccurate-with-a-calculator/

https://forum.talonro.com/bugtracker/gamebugreports/inaccurate-bowling-bash-damage-using-a-bow-r513/

A rogue with 99+17 dex + other atk gear = 432 atk,
using rental rudra + silver arrow, calc showing him 8.7~9k which is still the case at the moment.
In actual combat the person report getting about ~6k damage
based on BB formula it should be 432 atk * 500% bb lv10 * 2 HIT * 150% Holy vs Shadow2 = 6480.

**Comment by nattwara** [comment]
Update : this isn't about element anymore. That 2 bows is bugged.
This bug still happen even with rudra + normal arrow vs poring.

**Comment by nattwara** [comment]
Update: damage also weird with LK BB.
Lv99 LK, 99+1 STR, 1+0 DEX, 1+0 LUK, eden 2h sword III, total 400 atk (200 from weapon).
BB Lv 10.
VS
bug testing monster, size = large, lv1, 1m HP, 0+0 def, 0+0 mdef, perfect hit = 21 Neutral 4.

Result:
Minimum Damage | 2010 (1005 x 2hit)
Average Damage | 3000 (1500 x 2hit)
Maximum Damage | 3990 (1995 x 2hit)

When change 1dex to 99dex, atk changed to 419
new result:
Minimum Damage | 3570 (1785 x 2hit)
Average Damage | 3874 (1937 x 2hit)
Maximum Damage | 4180 (2090 x 2hit)

Found this n_A_DMG array with these value during testing. Can't find where it was declared.

1dex:
n_A_DMG have value of [201, 300, 399]
after calculate BB skill, [1005, 1500, 1995]

99dex:
n_A_DMG have value of [357, 387.5, 418]
after calculate BB skill, [1785, 1937, 2090]

**Comment by nattwara** [comment]
Update:
I made more bow for testing,
A neutral bow with 50 ATK and another one with 500ATK

When I use low ATK bow and use 4 inca card, the damage is accurate.
This same bow with high dex also have accurate damage.

But with 500ATK bow, with low dex is it a little bit off, but close to being accurate.
With higher dex, 98+12, it damage is way off. (734 ATK shown in calc, 11k damage)

And I added 50,000 ATK in manual edit then retest in all case,
I got 5xx,xxx where the 500k damage come from ATK of manual edit, and the xx,xxx is exactly the same as if adding damage value from previous test.
Ex: 500ATK bow 98+12DEX, 50k ATK from manual edit = 511k

It's something to do with the bow having high ATK and that got some weird calculation from the formula involving DEX, where higher dex = closer to actual formula value.

**Comment by nattwara** [comment]
Bowling Bash Exponentially Exploded.

Pic 1 without gear

![chrome_2018-07-13_15-18-28](https://user-images.githubusercontent.com/3313046/42680965-a9d86326-86b0-11e8-9308-fd3d6a9a310d.png)


Pic 2 with gear

![chrome_2018-07-13_15-18-11](https://user-images.githubusercontent.com/3313046/42680971-adcefcf6-86b0-11e8-8855-be75c1a8878f.png)


Pic 3-5 with manual edit

![chrome_2018-07-13_15-19-21](https://user-images.githubusercontent.com/3313046/42680981-b26751e6-86b0-11e8-86e6-8add1497684e.png)

![chrome_2018-07-13_15-19-31](https://user-images.githubusercontent.com/3313046/42680990-b5907aaa-86b0-11e8-9074-ca4fdfc0e8fa.png)

![chrome_2018-07-13_15-20-02](https://user-images.githubusercontent.com/3313046/42680997-b8745d40-86b0-11e8-98f9-5ae9b290b291.png)

**Comment by nattwara** [comment]
What should actually be expected from BB Lv10:
Ps: This test is neutral attack vs neutral monster.

DEX give ATK for bow, DEX make damage closer to ideal value.
(formula value, if we were to calculate ourselves)
If we let the ATK from DEX, cards, manual, Bow, everything = n_ATK

With very high dex we should expect to see minimum, average, and maximum being close to one another with the value of n_ATK * 500% * 2 Hit

During the testing with LK, the ideal value is higher than that of maximum value,
and the max value being close to ideal value, with the avg and min being noticeably lower.
And with 0+1 DEX, the min get 0.5x of ideal value, and avg get 0.75 of ideal value.
This is a pic of LK with 1+9999 DEX, 99+1 STR, Eden 2H sword III for total of 2400 ATK
VS the test monster, large size variant:

![chrome_2018-07-13_15-29-56](https://user-images.githubusercontent.com/3313046/42681385-b9c090fa-86b1-11e8-9eec-8d2b39db11c3.png)

which conform to the theory about DEX.
(which was taken from multiple source including talonro wiki and iro classic wiki about stats)

So this should mean that, that function that do the dex calculation part for getting close to ideal value should return something between 0.5 and 1 for multiplying with the ideal value.
(0.5-1 for min, 0.75-1 for avg, and something-1 for max)

Ex: 1000 ATK BB Lv 10
If your current dex give:
max = 0.97 x ideal value
avg = 0.91 x ideal value
min = 0.85 x ideal value.
Then the damage should be:
max = 970 x 500% x 2 HIT
avg = 910 x 500% x 2 HIT
min = 850 x 500% x 2 HIT

But the bow test show having very high dex that should make it get close to ideal value, when having 60k ATK, which for BB Lv10, should have been 300k x 2HIT = 600k, but instead, we have over 8 millions

**Comment by nattwara** [comment]
Now the only workaround I can think of is set all bow atk to 0, put the attack at the end of bow script, 17,bow_atk, like Incantation Samurai Card. And use foot.js/head.js to detect if it's a bow, set the ATK in the item info box to the bow's atk. As in the picture:

![chrome_2018-07-13_15-54-18](https://user-images.githubusercontent.com/3313046/42682560-1f5eb876-86b5-11e8-9573-17413fe1cdf3.png)

(tbh; this workaround look ugly and probably will make any future person who work on this hate it. lol)

Otherwise we'll have to figure out the math thing behind this.

**Comment by nattwara** [comment]
Summary of my assumption:
Pre-requiste: 
element modifier is not taken into account (neutral vs neutral)
monster size vs weapon is not taken into account. (2h sword vs large = 100%, bow vs small = 100%)

1) Apart from damage for bow type weapon, dex make damage get close to ideal value
2) min damage is between 0.5x - 1x of ideal value
3) avg damage is between 0.75x-1x
4) max damage is very close to ideal value.
5) with very high dex, and if the resulting atk is 1,000. Lv10 BB should deal 5,000 x 2 HIT for ideal value.

**Comment by TensaiKato** [comment]
Yes, it seems very accurate, the min, max and avg are close damage and testing with 1000 ATK should give 5000 x2.

I wouldn't change the bow attack (not even only in calculation), maybe apply the x * the modifiers you said?

**Comment by nattwara** [comment]
Anddd.. it's fixed 😆

There was a formula that check for weapon type then re-adjust the damage for ranged weapons. Bow & all the guns.

And if so, there is this ..... Weapon_ATK * Weapon_ATK / 100 ..... so that's why it exponentially exploded when I tried making bow with 50k atk, anything above 100 also give weird damage, the higher the damage, the weirder it get and nothing seem to be happening with low atk bow.

Added check if it's Bow+BB then don't re-adjust because that is considered a melee.
Now it should behave the same thing as LK BB.
